src/VisualStudio/Core/Def/Implementation/CodeLens/RemoteCodeLensReferencesService.cs

```diff
@@ -104,11 +97,135 @@ internal sealed class RemoteCodeLensReferencesService : ICodeLensReferencesServi
+                }
+
+                var span = new TextSpan(descriptor.SpanStart, descriptor.SpanLength);
+                var results = await spanMapper.MapSpansAsync(document, SpecializedCollections.SingletonEnumerable(span), cancellationToken).ConfigureAwait(false);
+
+                // external component violated contracts. the mapper should preserve input order/count. 
+                // since we gave in 1 span, it should return 1 span back
+                Contract.ThrowIfTrue(results.IsDefaultOrEmpty);
+
+                var result = results[0];
+                if (result.IsDefault)
+                {
+                    // it is allowed for mapper to return default 
+                    // if it can't map the given span to any usable span
+                    continue;
+                }
+
+                var excerpter = document.Services.GetService<IDocumentExcerptService>();
+                var referenceExcerpt = await excerpter.TryExcerptAsync(document, span, ExcerptMode.SingleLine, cancellationToken).ConfigureAwait(false);
+                var tooltipExcerpt = await excerpter.TryExcerptAsync(document, span, ExcerptMode.Tooltip, cancellationToken).ConfigureAwait(false);
+
+                var reference = GetReferenceInfo(referenceExcerpt, descriptor);
+                var referenceTexts = GetReferenceTexts(referenceExcerpt, tooltipExcerpt, descriptor);
+
+                list.Add(new ReferenceLocationDescriptor(
+                    descriptor.LongDescription,
+                    descriptor.Language,
+                    descriptor.Glyph,
+                    result.Span.Start,
+                    result.Span.Length,
+                    result.LinePositionSpan.Start.Line,
+                    result.LinePositionSpan.Start.Character,
+                    descriptor.ProjectGuid,
+                    descriptor.DocumentGuid,
+                    result.FilePath,
+                    reference.text,
+                    reference.start,
+                    reference.length,
+                    referenceTexts.before1,
+                    referenceTexts.before2,
+                    referenceTexts.after1,
+                    referenceTexts.after2));
+            }
+
+            return list;
+        }
+
+        private static (string text, int start, int length) GetReferenceInfo(ExcerptResult? reference, ReferenceLocationDescriptor descriptor)
+        {
+            if (reference.HasValue)
+            {
+                return (reference.Value.Content.ToString().TrimEnd(),
+                        reference.Value.MappedSpan.Start,
+                        reference.Value.MappedSpan.Length);
+            }
+
+            return (descriptor.ReferenceLineText, descriptor.ReferenceStart, descriptor.ReferenceLength);
+        }
+
+        private static (string before1, string before2, string after1, string after2) GetReferenceTexts(ExcerptResult? reference, ExcerptResult? tooltip, ReferenceLocationDescriptor descriptor)
```

Comment:
before1/2/after1/2 is really vague.  i don't really have any concept of what's going on here.

Reply:
before and after is text before and after line that contains codelens span. why designed this way. I don't know, but this is how it is written, so I am keeping it as it is.

no plan to change in this PR since that will bereak VS integration.

Reply:
So these static methods are called directly by VS?  Are they using reflection to do this?  If so, we need to mark these methods in some way as having signatures we cannot change.

Reply:
not talking about the method, talking about before1, before2, after1, after2. the shape of ReferenceLocationDescriptor

Reply:
ok.  but this question is about this method :)   If this method is something we own, can we use naming that makes things clear?